Common/TableProducer/OccupancyTableProducer.cxx

```diff
@@ -0,0 +1,1159 @@
+
+#include <typeinfo>
+
+#include "Framework/AnalysisDataModel.h"
+#include "Framework/HistogramRegistry.h"
+#include "CCDB/BasicCCDBManager.h"
+#include "DataFormatsFT0/Digit.h"
+#include "DataFormatsParameters/GRPLHCIFData.h"
+// #include "MetadataHelper.h"
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+
+int32_t nBCsPerOrbit = o2::constants::lhc::LHCMaxBunches;
+
+struct occTableProducer {
+
+  Service<o2::ccdb::BasicCCDBManager> ccdb;
+
+  // declare production of tables
+  Produces<aod::OccIndexTable> GenOccIndexTable;
+  Produces<aod::Occs> GenOccTable;
+
+  Configurable<int> customOrbitOffset{"customOrbitOffset", 0, "customOrbitOffset for MC"};
+
+  // Histogram registry;
+  //  HistogramRegistry recoTracks  {"recoTracks"  , {}, OutputObjHandlingPolicy::AnalysisObject, true, true};
+  HistogramRegistry recoEvent{"recoEvent", {}, OutputObjHandlingPolicy::AnalysisObject, true, true};
+
+  void init(InitContext const&)
+  {
+    LOGF(info, "Starting init");
+
+    // Getting Info from CCDB, to be implemented
+    //  auto& mgr = o2::ccdb::BasicCCDBManager::instance();
+    //  // http://ccdb-test.cern.ch:8080/browse/Users/r/raverma?report=true
+    //  mgr.setURL("http://ccdb-test.cern.ch:8080");
+    //  mgr.setCaching(true);
+    //  auto myVect = mgr.get<TVectorT<float>>("Users/r/raverma/");
+    //  if( !myVect){
+    //    LOG(info)<<"ERROR ::CCDB OBJECT Not found";
+    //    return;
+    //  }
+
+    recoEvent.add("h_nBCsPerTF", "h_nBCsPerTF", {HistType::kTH1F, {{100, 114040, 114060}}}); // 114048
+    recoEvent.add("h_nBCinTF", "h_nBCinTF", {HistType::kTH1F, {{1000, 0, 1000000}}});
+    recoEvent.add("h_RO_T0V0Prim_Unfm_80", "h_RO_T0V0Prim_Unfm_80:median contributors", {HistType::kTH1F, {{12 * 2, -1, 11}}});
+    recoEvent.add("h_RO_FDDT0V0Prim_Unfm_80", "h_RO_FDDT0V0Prim_Unfm_80:median contributors", {HistType::kTH1F, {{12 * 2, -1, 11}}});
+    recoEvent.add("h_RO_NtrackDet_Unfm_80", "h_RO_NtrackDetITS/TPC/TRD/TOF_80:median contributors", {HistType::kTH1F, {{12 * 2, -1, 11}}});
+
+    LOG(info) << "Printing Event Info ";
+    recoEvent.print();
+    LOG(info) << "Finishing Init ";
+  }
+
+  template <typename T, typename U>
+  int BinarySearchTable(T Key, U Table, int low, int high)
+  {
+    while (low <= high) {
+      int mid = low + (high - low) / 2;
+      if (Key == Table.iteratorAt(mid).trackId()) {
+        return mid;
+      }
+
+      if (Key > Table.iteratorAt(mid).trackId()) {
+        low = mid + 1;
+      } // If Key is greater, ignore left  half, update the low
+      else {
+        high = mid - 1;
+      } // If Key is smaller, ignore right half, update the high
+    }
+    return -1; // Element is not present
+  }
+
+  template <typename T, typename U>
+  int FindInTable(T key, U Table)
+  {
+    return BinarySearchTable(key, Table, 0, Table.size() - 1);
+  }
+
+  void FillNewListFromOldList(std::vector<int64_t>& NewList, std::vector<int64_t> OldList)
```

Comment:
There vector operations look like they can be implemented with standard algorithms, `std::sort`, `std::unique`, `std::transform`, etc. Please consider using them.

Reply:
I have implemented these changes, thanks